File: DisasterServer/Entities/VVLava.cs
using DisasterServer.Maps;
using DisasterServer.Session;
using DisasterServer.State;

namespace DisasterServer.Entities;

public class VVLava : Entity
{
	public float StartY;

	public float TravelDistance;

	public byte ID;

	private byte _state;

	private int _timer = 1200;

	private float _y;

	private float _accel;

	private static Random _rand = new Random();

	public VVLava(byte id, float startY, float dist)
	{
		ID = id;
		StartY = startY;
		_y = startY;
		TravelDistance = dist;
	}

	public override TcpPacket? Destroy(Server server, Game game, Map map)
	{
		return null;
	}

	public override TcpPacket? Spawn(Server server, Game game, Map map)
	{
		_timer += _rand.Next(2, 5) * 60;
		return null;
	}

	public override UdpPacket? Tick(Server server, Game game, Map map)
	{
		switch (_state)
		{
		case 0:
			_y = StartY + MathF.Sin((float)_timer / 25f) * 6f;
			if (_timer-- <= 0)
			{
				_state = 1;
			}
			break;
		case 1:
			if (_y < StartY + 20f)
			{
				_y += 0.15f;
			}
			else
			{
				_state = 2;
			}
			break;
		case 2:
			if (_y > StartY - TravelDistance)
			{
				if (_accel < 5f)
				{
					_accel += 0.08f;
				}
				_y -= _accel;
			}
			else
			{
				_state = 3;
				_timer = 300;
				_accel = 0f;
			}
			break;
		case 3:
			if (_timer-- <= 0)
			{
				_state = 4;
			}
			_y = StartY - TravelDistance + MathF.Sin((float)_timer / 25f) * 6f;
			break;
		case 4:
			if (StartY > _y)
			{
				if (_accel < 5f)
				{
					_accel += 0.08f;
				}
				_y += _accel;
			}
			else
			{
				_state = 0;
				_timer = 1200;
				_accel = 0f;
			}
			break;
		}
		return new UdpPacket(PacketType.SERVER_VVLCOLUMN_STATE, ID, _state, _y);
	}
}

File: disasterserver/entities/VVLava.c
#include <entities/VVLava.h>
#include <CMath.h>

bool lava_tick(Server* server, Entity* entity)
{
	Lava* lv = (Lava*)entity;
	switch (lv->state)
	{
		case LV_IDLE: // Move using sin
		{
			lv->pos.y = lv->start + sinf((float)lv->timer / 25.0f) * 6;

			lv->timer -= server->delta;
			if (lv->timer <= 0)
				lv->state = LV_MOVEDOWN;

			break;
		}

		case LV_MOVEDOWN: // Move down a bit
		{
			if (lv->pos.y < lv->start + 20)
			{
				lv->pos.y += 0.15f * (float)server->delta;
			}
			else
				lv->state = LV_RAISE;

			break;
		}

		case LV_RAISE: // accelerate
		{
			if (lv->pos.y > lv->start - lv->dist)
			{
				lv->pos.y -= lv->vel;

				if (lv->vel < 5)
					lv->vel += 0.08f * (float)server->delta;
				else
					lv->vel = 5;
			}
			else
			{
				lv->state = LV_MOVE;
				lv->timer = (4.0 + rand() % 3) * TICKSPERSEC;
				lv->vel = 0;
			}

			break;
		}

		case LV_MOVE: // move on spot
		{
			lv->pos.y = (lv->start - lv->dist) + sinf((float)lv->timer / 25.0f) * 6;

			lv->timer -= server->delta;
			if (lv->timer <= 0)
				lv->state = LV_LOWER;
			break;
		}

		case LV_LOWER: // go down
		{
			if (lv->start > lv->pos.y)
			{
				lv->pos.y += lv->vel * (float)server->delta;

				if (lv->vel < 5)
					lv->vel += 0.08f * (float)server->delta;
				else
					lv->vel = 5;
			}
			else
			{
				lv->state = LV_IDLE;
				lv->timer = (20.0 + rand() % 5) * TICKSPERSEC;
				lv->vel = 0;
			}

			break;
		}
	}

	Packet pack;
	PacketCreate(&pack, SERVER_VVLCOLUMN_STATE);
	PacketWrite(&pack, packet_write8, lv->lid);
	PacketWrite(&pack, packet_write8, lv->state);
	PacketWrite(&pack, packet_writefloat, lv->pos.y);
	server_broadcast(server, &pack, false);

	return true;
}
